### PR TITLE
[DataFrame] Allow mixed type concat

### PR DESCRIPTION
**Pull request by adgirish**
Allowing for correct mixed type concats,

Corrected mutation of Pandas dataframe when from_pandas called

Added tests for inner join and None values in joining tables

## Related issue number

#1865

### CODE REVIEW

File: python/ray/dataframe/concat.py [hunk 1]
```diff
@@ -21,45 +21,38 @@ def _concat(frame1, frame2):
...
-                  "Obj as dicts not implemented. To contribute to "
-                  "Pandas on Ray, please visit github.com/ray-project/ray."
-                  )
-
+        f1_columns, f2_columns = frame1.columns, frame2.columns
+        
         # Case 2: Both are different types
         if isinstance(frame1, pd.DataFrame):
             frame1 = from_pandas(frame1, len(frame1) / 2**16 + 1)
+            frame1.columns = f1_columns
```
Comment (kunalgosar) [inline python/ray/dataframe/concat.py]: from_pandas already sets the columns in the output df

**Comment by adgirish** [comment]
Jenkins, retest this please.

**Review by kunalgosar — CHANGES_REQUESTED** [review]

File: python/ray/dataframe/concat.py (hunk 1 above)
Comment (adgirish) [reply python/ray/dataframe/concat.py]: They aren't being set properly though when converting a pd.DataFrame in the mixed frame tests. Use the df, df2, df3 test and put a print after to verify. This is where the `0, 1, 2, 4` column names are coming from

Comment (adgirish) [reply python/ray/dataframe/concat.py]: Resolved

**Review by kunalgosar — APPROVED** [review]

**Comment by adgirish** [comment]
Jenkins, retest this please.

File: python/ray/dataframe/utils.py [hunk 2]
```diff
@@ -91,13 +91,14 @@ def from_pandas(df, num_partitions=None, chunksize=None):
         A new Ray DataFrame object.
     """
     from .dataframe import DataFrame
+    df_op = df.copy(deep=True)
```
Comment (devin-petersohn) [inline python/ray/dataframe/utils.py]: Why is the copy needed here?

Comment (adgirish) [reply python/ray/dataframe/utils.py]: You're right, we no longer need this

**Comment by adgirish** [comment]
Jenkins, retest this please.

**Comment by adgirish** [comment]
Jenkins, retest this please.

**Comment by adgirish** [comment]
Jenkins, retest this please.

**Comment by devin-petersohn** [comment]
Jenkins, retest this please.

**Comment by devin-petersohn** [comment]
Please rebase on master to clear the Jenkins error

**Comment by adgirish** [comment]
Jenkins, retest this please.

**Comment by adgirish** [comment]
Jenkins, retest this please.

**Comment by devin-petersohn** [comment]
Jenkins, retest this please.

**Comment by adgirish** [comment]
Jenkins, retest this please.

**Comment by devin-petersohn** [comment]
Jenkins, retest this please.

**Comment by adgirish** [comment]
Jenkins, retest this please.

**Review by devin-petersohn — APPROVED** [review]

**Comment by devin-petersohn** [comment]
#1932 also fixes this error and adds new functionality. Given that the code structure is different now, I think it best to close without merging. @adgirish is this ok?

**Comment by adgirish** [comment]
Sounds good!

**Comment by devin-petersohn** [comment]
Closing in favor of #1932.

**Pull request closed without merging**